File: cpp_drill/Problems/STK03_GenerateParentheses.cpp
// REF : https://leetcode.com/problems/generate-parentheses/

#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <algorithm>
#include <cstring>
using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        stack<char> st;
        backtrack(0, 0, n, st, res);
        return res;
    }

    void backtrack(int open_cnt, int close_cnt, int &n, stack<char> &st, vector<string> &res) {
        if (open_cnt == close_cnt and close_cnt == n) {
            string ans;
            for (stack<char> temp_st = st; !temp_st.empty(); temp_st.pop()) {
                ans.push_back(temp_st.top());
            }
            string rev = string(ans.rbegin(), ans.rend());
            res.push_back(rev);
            return;
        }

        if (open_cnt < n) {
            st.push('(');
            backtrack(open_cnt + 1, close_cnt, n, st, res);
            st.pop();
        }

        if (close_cnt < open_cnt) {
            st.push(')');
            backtrack(open_cnt, close_cnt + 1, n, st, res);
            st.pop();
        }
    }

};


class Solution2 {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        addingpar(res, "", n, 0);
        return res;
    }
    void addingpar(vector<string> &v, string str, int n, int m){
        if(n==0 && m==0) {
            v.push_back(str);
            return;
        }

        if (n > 0)
            addingpar(v, str+"(", n-1, m+1); 

        if (m > 0)
            addingpar(v, str+")", n, m-1);
    }
};


int main() {

    // vector<string> out = Solution().generateParenthesis(3);
    vector<string> out = Solution2().generateParenthesis(3);
    cout << "output : " << endl;
    for (auto &i : out) {
        cout << i << endl;
    }
    return 0;
}

File: py_drill/Problems/STK03_GenerateParentheses.py
# REF : https://leetcode.com/problems/generate-parentheses/
# NOTES :
#       Stack + DFS + Backtracking => Solution 1
#       if open count == close count == n then it's one of the results
#       if open count < n then we can add "("
#       if close count < open count then we can add ")"

#       DFS => Solution 2
#       Use two integers to count the remaining left parenthesis (n)
#       and the right parenthesis (m) to be added.
#       At each function call add a left parenthesis if n > 0
#       and add a right parenthesis if m > 0

# T: O( 4^n / n * sqrt(n) )
# S: O( 4^n / n * sqrt(n) )
class Solution:
    def generateParenthesis(self, n: int) -> list[str]:

        res = []
        stack = []

        def backtrack(open_cnt, close_cnt):
            if open_cnt == close_cnt == n:
                res.append("".join(stack))
                return

            if open_cnt < n:
                stack.append("(")
                backtrack(open_cnt + 1, close_cnt)
                stack.pop()     # Need to pop the same element that we added

            if close_cnt < open_cnt:
                stack.append(")")
                backtrack(open_cnt, close_cnt + 1)
                stack.pop()     # Need to pop the same element that we added

        backtrack(0, 0)
        return res

# T: O(2^n)
# S: O(2^n)
class Solution2:
    def generateParenthesis(self, n: int) -> list[str]:
        res = []

        def dfs(n, m, ans):
            if n == 0 and m == 0:
                res.append(ans)
                return

            if n > 0:
                dfs(n-1, m+1, ans + '(')
            if m > 0:
                dfs(n, m-1, ans + ')')

        dfs(n, 0, "")
        return res

def main():
    print(Solution().generateParenthesis(3))
    print(Solution2().generateParenthesis(3))

if __name__ == "__main__":
    main()